driver-core/src/main/com/mongodb/internal/connection/LoadBalancedCluster.java

```diff
@@ -181,50 +202,211 @@ public ServerTuple selectServer(final ServerSelector serverSelector) {
+        }
+        return settings.getServerSelectionTimeout(NANOSECONDS);
+    }
+
+    private void notifyWaitQueueHandler(final ServerSelectionRequest request) {
+        lock.lock();
+        try {
+            if (isClosed()) {
+                request.onError(createShutdownException());
+                return;
+            }
+            if (initializationCompleted) {
+                handleServerSelectionRequest(request);
+                return;
+            }
+
+            waitQueue.add(request);
+
+            if (waitQueueHandler == null) {
+                waitQueueHandler = new Thread(new WaitQueueHandler(), "cluster-" + clusterId.getValue());
+                waitQueueHandler.setDaemon(true);
+                waitQueueHandler.start();
+            } else {
+                condition.signalAll();
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private final class WaitQueueHandler implements Runnable {
+        public void run() {
+            List<ServerSelectionRequest> timeoutList = new ArrayList<>();
+            while (!(isClosed() || initializationCompleted)) {
+                lock.lock();
+                try {
+                    if (isClosed() || initializationCompleted) {
+                        break;
+                    }
+                    long waitTimeNanos = Long.MAX_VALUE;
+                    long curTimeNanos = System.nanoTime();
+
+                    for (Iterator<ServerSelectionRequest> iterator = waitQueue.iterator(); iterator.hasNext();) {
+                        ServerSelectionRequest next = iterator.next();
+                        long remainingTime = next.getRemainingTime(curTimeNanos);
+                        if (remainingTime <= 0) {
+                            timeoutList.add(next);
+                            iterator.remove();
+                        } else {
+                            waitTimeNanos = Math.min(remainingTime, waitTimeNanos);
+                        }
+                    }
+                    if (timeoutList.isEmpty()) {
+                        try {
+                            //noinspection ResultOfMethodCallIgnored
+                            condition.await(waitTimeNanos, NANOSECONDS);
+                        } catch (InterruptedException e) {
+                            fail();
+                        }
+                    }
+                } finally {
+                    lock.unlock();
+                }
+
+                timeoutList.forEach(request -> request.onError(createTimeoutException()));
+                timeoutList.clear();
+            }
+
+            List<ServerSelectionRequest> shutdownList;
+            lock.lock();
+            try {
+                shutdownList = new ArrayList<>(waitQueue);
+                waitQueue.clear();
+            } finally {
+                lock.unlock();
+            }
+            shutdownList.forEach(request -> request.onError(createTimeoutException()));
```

Comment:
This code is executed either after closing the `LoadBalancedCluster` or after initializing it. In the latter case, the queue is guaranteed to be empty (`DnsSrvRecordInitializer.initialize`clears it and no thread adds new elements to it), so `shutdownList` is not empty iff `LoadBalancedCluster` is closed.

If the above understanding is correct, shouldn't callbacks in `shutdownList` be completed with `createShutdownException`?

Also, I think it would be useful to leave a code comment explaining why the code around `shutdownList` is needed.

Reply:
Agreed. Done.